# [\#8252 Issue](https://github.com/ethereum/solidity/issues/8252) `closed`: Add internal revert reasons to Yul codegen

#### <img src="https://avatars.githubusercontent.com/u/504195?u=ce2facd14af9fd474ebff49f0d44891f56f7500f&v=4" width="50">[leonardoalt](https://github.com/leonardoalt) opened issue at [2020-02-05 23:35](https://github.com/ethereum/solidity/issues/8252):

Similar to https://github.com/ethereum/solidity/pull/8178

- [ ] `codegen/YulUtilFunctions.cpp`
- [ ] `codegen/ir/*`
- [ ] tests
- [ ] docs

#### <img src="https://avatars.githubusercontent.com/u/9073706?v=4" width="50">[chriseth](https://github.com/chriseth) commented at [2020-02-05 23:35](https://github.com/ethereum/solidity/issues/8252#issuecomment-603977359):

Some of these should actually be invalid opcodes.

#### <img src="https://avatars.githubusercontent.com/u/20340?v=4" width="50">[axic](https://github.com/axic) commented at [2020-02-05 23:35](https://github.com/ethereum/solidity/issues/8252#issuecomment-734332617):

Current cases:

```
semanticTests/revertStrings/unknown_sig_no_fallback.sol: FAIL
  Contract:
    contract A {
    	receive () external payable {}
    }
    // ====
    // EVMVersion: >=byzantium
    // revertStrings: debug

  Running via Yul:
  Expected result:
  // (): hex"00" -> FAILUREhex"08c379a0", 0x20, 41, "Unknown signature and no fallbac", "k defined"

  Obtained result:
  // (): hex"00" -> FAILURE
  Warning: The call to "()" returned 
```

```
semanticTests/revertStrings/ether_non_payable_function.sol: FAIL
  Contract:
    contract C {
    	function f() public {}
    }
    // ====
    // EVMVersion: >=byzantium
    // revertStrings: debug

  Running via Yul:
  Expected result:
  // f(), 1 ether -> FAILUREhex"08c379a0", 0x20, 34, "Ether sent to non-payable functi", "on"
  // () -> FAILUREhex"08c379a0", 0x20, 53, "Contract does not have fallback ", "nor receive functions"

  Obtained result:
  // f(), 1 ether -> FAILURE
  Warning: The call to "f()" returned 

  // () -> FAILURE
  Warning: The call to "()" returned 
```

```
semanticTests/revertStrings/enum.sol: FAIL
  Contract:
    contract C {
    	enum E {X, Y}
    	function f(E[] calldata arr) external {
    		arr[1];
    	}
    }
    // ====
    // ABIEncoderV1Only: true
    // EVMVersion: >=byzantium
    // revertStrings: debug

  Running via Yul:
  Expected result:
  // f(uint8[]): 0x20, 2, 3, 3 -> FAILUREhex"08c379a0", 0x20, 17, "Enum out of range"

  Obtained result:
  // f(uint8[]): 0x20, 2, 3, 3 -> FAILURE
  Warning: The call to "f(uint8[])" returned 
```

```
semanticTests/revertStrings/called_contract_has_code.sol: FAIL
  Contract:
    contract C {
    	function f() external {}
    	function g() external {
    		C c = C(0x0000000000000000000000000000000000000000000000000000000000000000);
    		c.f();
    	}
    }
    // ====
    // EVMVersion: >=byzantium
    // revertStrings: debug

  Running via Yul:
  Expected result:
  // g() -> FAILUREhex"08c379a0", 0x20, 37, "Target contract does not contain", " code"

  Obtained result:
  // g() -> FAILURE
  Warning: The call to "g()" returned 
```

#### <img src="https://avatars.githubusercontent.com/u/20340?v=4" width="50">[axic](https://github.com/axic) commented at [2020-02-05 23:35](https://github.com/ethereum/solidity/issues/8252#issuecomment-734972320):

Actually for these cases I listed above shouldn't we use panic codes instead?

And if we do, we should change them in 0.8.0 for the old codegen too.

#### <img src="https://avatars.githubusercontent.com/u/9073706?v=4" width="50">[chriseth](https://github.com/chriseth) commented at [2020-02-05 23:35](https://github.com/ethereum/solidity/issues/8252#issuecomment-735710761):

https://github.com/ethereum/solidity/issues/8252#issuecomment-603977359

#### <img src="https://avatars.githubusercontent.com/u/20340?v=4" width="50">[axic](https://github.com/axic) commented at [2020-02-05 23:35](https://github.com/ethereum/solidity/issues/8252#issuecomment-735740608):

I'm not sure what you mean. Do you prefer `invalid()` now over panic codes?

Btw I added "Target contract does not contain code" and it is a one liner, so we could to the match-old-codegen if needed.

#### <img src="https://avatars.githubusercontent.com/u/9073706?v=4" width="50">[chriseth](https://github.com/chriseth) commented at [2020-02-05 23:35](https://github.com/ethereum/solidity/issues/8252#issuecomment-736553165):

Of course: old invalid -> new panic. But some of them should not be reverts.

#### <img src="https://avatars.githubusercontent.com/u/9073706?v=4" width="50">[chriseth](https://github.com/chriseth) commented at [2020-02-05 23:35](https://github.com/ethereum/solidity/issues/8252#issuecomment-736553270):

Error reverts I mean.


-------------------------------------------------------------------------------



[Export of Github issue for [ethereum/solidity](https://github.com/ethereum/solidity). Generated on 2022.05.23 at 03:51:38.]
